components/common/src/command/package/download.rs

```diff
@@ -0,0 +1,400 @@
+        {
+            return Err(Error::DownloadFailed(format!("We tried {} times but \
+                                                      could not download {} for \
+                                                      {}. Last error was: {}",
+                                                     RETRIES, ident, target, err)));
+        }
+
+        // At this point the artifact is in the cache...
+        let mut artifact = PackageArchive::new(self.cached_artifact_path(ident, target));
+        self.verify_artifact(ui, ident, target, &mut artifact)?;
+        Ok(artifact)
+    }
+
+    // This function and it's sibling in install.rs deserve to be refactored to eke out commonality.
+    /// Retrieve the identified package from the depot, ensuring that
+    /// the artifact is cached locally.
+    fn fetch_artifact<T>(&self,
+                         ui: &mut T,
+                         ident: &PackageIdent,
+                         target: PackageTarget)
+                         -> Result<()>
+        where T: UIWriter
+    {
+        ui.status(Status::Downloading, format!("{} for {}", ident, target))?;
+        match self.api_client.fetch_package((ident, target),
+                                            self.token,
+                                            self.artifact_download_path,
+                                            ui.progress())
+        {
+            Ok(_) => Ok(()),
+            Err(api_client::Error::APIError(StatusCode::NOT_IMPLEMENTED, _)) => {
+                println!("Host platform or architecture not supported by the targeted depot; \
+                          skipping.");
+                Ok(())
+            }
+            Err(e) => Err(Error::from(e)),
+        }
+    }
+
+    fn fetch_origin_key<T>(&self,
+                           ui: &mut T,
+                           name_with_rev: &str,
+                           token: Option<&str>)
+                           -> Result<()>
+        where T: UIWriter
+    {
+        ui.status(Status::Downloading,
+                  format!("{} public origin key", &name_with_rev))?;
+        let (name, rev) = parse_name_with_rev(&name_with_rev)?;
+        self.api_client.fetch_origin_key(&name,
+                                          &rev,
+                                          token,
+                                          self.key_download_path,
+                                          ui.progress())?;
+        ui.status(Status::Cached,
+                  format!("{} public origin key", &name_with_rev))?;
+        Ok(())
+    }
+
+    fn verify_artifact<T>(&self,
+                          ui: &mut T,
+                          ident: &PackageIdent,
+                          target: PackageTarget,
+                          artifact: &mut PackageArchive)
+                          -> Result<()>
+        where T: UIWriter
+    {
+        let artifact_ident = artifact.ident()?;
```

Comment:
We should not use any archive functions like ident() or target() that require opening and random accessing the package stream - it is very expensive